builtin/v10/evm/invariants.go

```diff
@@ -0,0 +1,36 @@
+package evm
+
+import (
+	"bytes"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v10/util/adt"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/xerrors"
+)
+
+// Checks internal invariants of account state.
+func CheckStateInvariants(st *State, store adt.Store) *builtin.MessageAccumulator {
+	acc := &builtin.MessageAccumulator{}
+
+	acc.Require(st.Nonce > 0, "EVM actor state nonce needs to be greater than 0")
+
+	byteCode, err := getBytecode(st.Bytecode, store)
+	acc.RequireNoError(err, "Unable to retrieve bytecode")
+
+	hasher := sha3.NewLegacyKeccak256()
+	hasher.Write(byteCode)
+	byteCodeHash := hasher.Sum(nil)
+
+	acc.Require(bytes.Equal(byteCodeHash, st.BytecodeHash[:]), "Bytecode hash doesn't match bytecode cid")
```

Comment:
Yeah, i think this is a good check to have.

Reply:
So this doesn't work at the moment hence the q before I spend more time on it.
I will need to modify the invariants API to have access to the state blockstore since that's where the bytecode resides. I'm not sure if thats fine or we try to avoid it since we only mainly try to look at the actor tree